## BUSA3020 - Assignment 1

---

**Assignment Points**: 100  
**Due Date**: Friday Week 7 (5 April 2024) @ 11.59pm  
**Instructions**: Provide answers within this Jupyter notebook using Python code, and submit via iLearn 
**Marking Criteria**: See end of document below   
**Assignment Background**: This assignment uses a dataset which is based on the Credit Card Defaults data discussed in Week 2 Tutorial  


---

**Problem 1** - Reading the dataset (Total Marks: 20)



**Q1**. Read the first 10,000 rows from the credit card dataset provided in the **assignment_data** folder 
- Name your DataFrame `df` 
- Delete `ID` column     

(5 marks) 




In [71]:
# ---- provide your code here -----

import pandas as pd
df = pd.read_excel('assignment_data/credit_data.xlsx', nrows = 10000)

del df['ID']

<hr style="width:25%;margin-left:0;"> 

**Q2**. List which *features* are *numeric*, *ordinal*, and *nominal* variables, and how many features of each kind there are in the dataset.
To answer this question 
- Find the definitions of the variables provided elsewhere in the course material (hint: make sure you do weekly tutorials)
- Find the definitions of numeric, ordinal and nominal variables
- Carefully consider the values of the data itself as well as the output of `df.info()`. 

Your answer should be written up in Markdown and include:
1) Definitions of the three kinds of variables,
2) A table listing all the features present in the dataset and their type (fill out the table template provided below) and
3) A brief description of the contents of the table.

|Variable Kind|Number of Features|Feature Names
| --- | --- | --- |
| Numeric | some number | some text |
| some text  | some number | some text |
| some text  | some number | some text |


(10 marks)

In [72]:
# ---- provide your code here -----

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 24 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   LIMIT_BAL                   10000 non-null  int64  
 1   SEX                         9617 non-null   float64
 2   EDUCATION                   9617 non-null   float64
 3   MARRIAGE                    9617 non-null   float64
 4   AGE                         9617 non-null   float64
 5   PAY_0                       9617 non-null   float64
 6   PAY_2                       9617 non-null   float64
 7   PAY_3                       9617 non-null   float64
 8   PAY_4                       9642 non-null   float64
 9   PAY_5                       9642 non-null   float64
 10  PAY_6                       9642 non-null   float64
 11  BILL_AMT1                   9642 non-null   float64
 12  BILL_AMT2                   9642 non-null   float64
 13  BILL_AMT3                   9642

---- provide your text answer here ----

- **Numeric Variables:** Quantitative values measured on a numerical level.
- **Ordinal Variables:** Categorical values which have a natural ordering or ranking. 
- **Nominal Variables:** Categorical values without any ordering.


|Variable Kind|Number of Features|Feature Names
| --- | --- | --- |
| Numeric | 14 | LIMIT_BAL, AGE, BILL_AMT1, BILL_AMT2, BILL_AMT3, BILL_AMT4, BILL_AMT5, BILL_AMT6, PAY_AMT1, PAY_AMT2, PAY_AMT3, PAY_AMT4, PAY_AMT5, PAY_AMT6|
| Ordinal  | 7 | EDUCATION, PAY_0, PAY_2, PAY_3, PAY_4, PAY_5, PAY_6 |
| Nominal | 3 | SEX, MARRIAGE, default payment next month|



The column **Variable Kind** describes the types of features in the dataset, which is categorized in three groups: Numeric, Ordinal, and Nominal. 

The column **Number of Features** indicates how many features belong to the each type. The results show that there are 14 numeric, 7 ordinal, and 3 nominal variables.  

Lastly, the column **Feature Names** lists the names of the features of each type.

- Numeric variables include LIMIT_BAL, AGE, BILL_AMT1 to 6, and PAY_AMT1 to 6
- Ordinal variables include EDUCATION, and PAY_0, 2 to 6
- Nominal variables include SEX, MARRIAGE, and default payment next month

<hr style="width:25%;margin-left:0;"> 

**Q3.** Missing Values. 

- Print out the number of missing values for each variable in the dataset and comment on your findings.

(5 marks)

In [73]:
# ---- provide your code here -----
df.isnull().sum()

LIMIT_BAL                       0
SEX                           383
EDUCATION                     383
MARRIAGE                      383
AGE                           383
PAY_0                         383
PAY_2                         383
PAY_3                         383
PAY_4                         358
PAY_5                         358
PAY_6                         358
BILL_AMT1                     358
BILL_AMT2                     358
BILL_AMT3                     358
BILL_AMT4                     360
BILL_AMT5                     360
BILL_AMT6                     360
PAY_AMT1                      360
PAY_AMT2                      360
PAY_AMT3                        2
PAY_AMT4                        2
PAY_AMT5                      300
PAY_AMT6                      300
default payment next month      0
dtype: int64

---- provide your text answer here ----

Tha dataset contains a large number of missing values, which are more noticeable in columns with 383 counts such as SEX, EDUCATION, MARRIAGE, AGE, PAY_0, PAY_2, and PAY_3. However, the features such as LIMIT_BAL, PAY_AMT3, PAY_AMT4, and default payment next month show minimal missing data, proving greater reliabilty. 

To address these missing values, the data imputation strategy could help to replace them with similar values instead of leaving the data incomplete.  

---
---

**Problem 2.** Cleaning data and dealing with categorical features (Total Marks: 40)


**Q1.** 

- Use an appropriate `pandas` function to impute missing values using one of the following two strategies: `mean` and `mode`. (10 marks)
    - Take into consideration the type of each variable  and the best practices we discussed in class/lecture notes
- Explain what data imputation is, how you have done it here, and what decisions you had to make. (5 marks)



(Total: 15 marks)

In [74]:
# ---- provide your code here -----
from sklearn.impute import SimpleImputer
import numpy as np

numeric = ['LIMIT_BAL', 'AGE', 'BILL_AMT1', 'BILL_AMT2', 'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'PAY_AMT1', 'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6']
ordinal = ['EDUCATION', 'PAY_0', 'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6']
nominal = ['SEX', 'MARRIAGE', 'default payment next month']

for column in numeric:
    df[column].fillna(df[column].mean(axis = 0), inplace=True)
    
for column in nominal + ordinal:
    df[column].fillna(df[column].mode()[0], inplace=True)

---- provide your text answer here ----

Data imputation is a method to deal with missing values in the dataset by replacing with estimated or calculated values. We can use different interpolation to fill in the missing observations, namely mean, median and mode. 

Here, I have to decide which interpolation I should use for different variables. 
- For numeric variables, I used **mean** since the distribution is symmetrical and no significant effect by outliers. 
- For the categorical variables like ordinal and nominal, I used **mode** as it gives the most frequent value, which is a better estimator of each feature. 

By replacing missing values with mean for numaric and mode for categorical features using `for`loop and `fillna` method, the data is now complete with no gaps in the data.

<hr style="width:25%;margin-left:0;"> 

**Q2**. 
- Print `value_counts()` of the 'SEX' column and add a dummy variable named 'SEX_FEMALE' to `df` using `get_dummies()` (3 marks)
- Carefully explain what the values of the new variable 'SEX_FEMALE' mean (2 mark)
- Make sure the variable 'SEX' is deleted from `df`

(Total: 5 marks)  

In [75]:
# ---- provide your code here -----
print(df['SEX'].value_counts())

df['SEX_FEMALE'] = pd.get_dummies(df['SEX'])[2.0]

del df['SEX']

SEX
2.0    6206
1.0    3794
Name: count, dtype: int64


---- provide your text answer here ----

The new variable 'SEX_FEMALE' is derived from the original 'SEX' variable, which represents the gender of individuals in the dataset. A value of 1.0 indicates male gender, whereas 2.0 indicates female. 

I tried to genereate 'SEX_FEMALE' variable using `get_dummies()`, so that a value of **2.0** denotes **true (female)** and **1.0** denotes **false (male)**. 


<hr style="width:25%;margin-left:0;"> 

**Q3**. Print `value_counts()` of the 'MARRIAGE' column and *carefully* comment on what you notice in relation to the definition of this variable. 

(Total: 5 marks) 

In [76]:
# ---- provide your code here -----
df['MARRIAGE'].value_counts()

MARRIAGE
2.0    5511
1.0    4361
3.0     110
0.0      18
Name: count, dtype: int64

---- provide your text answer here ----

The variable 'MARRIAGE' represents different **marital status** based on these nominal values.

- The value **1.0** represents **married** individuals, with **4537** counts.
- The value **2.0** represents **single** individuals, with **5677** counts.
- The value **3.0** represents **unknown** individuals, with **113** counts.
- The value **0.0** represents **lack of data** about marital status, with **18** counts.

The value **0.0** in the 'MARRIAGE' columnn, indicating a lack of information regarding marital status, could potentially lower the accuracy of the dataset. Thus, we need to consider alternative options to replace those values.

<hr style="width:25%;margin-left:0;"> 

**Q4**. 

- Apply `get_dummies()` to 'MARRIAGE' feature and add dummy variables 'MARRIAGE_MARRIED', 'MARRIAGE_SINGLE', 'MARRIAGE_OTHER' to `df`. (5 marks)   
- Carefully consider how to allocate all the values of 'MARRIAGE' across these 3 newly created features (5 marks)   
    - Explain what decisions you had to make
- Make sure that 'MARRIAGE' is deleted from `df`   

(Total: 10 marks)   

In [77]:
# ---- provide your code here -----
df['MARRIAGE'].replace({0.0: df['MARRIAGE'].mode()[0]}, inplace=True)

df['MARRIAGE_MARRIED'] = pd.get_dummies(df['MARRIAGE'])[1.0]
df['MARRIAGE_SINGLE'] = pd.get_dummies(df['MARRIAGE'])[2.0]
df['MARRIAGE_OTHER'] = pd.get_dummies(df['MARRIAGE'])[3.0]

del df['MARRIAGE']

---- provide your text answer here ----

I replaced 0.0 with the value **2.0** which is the most frequent value in the dataset, assuming that people with missing information has the marital status similar to single individuals (2.0). 

Afterwards, I created three new columns: 'MARRIAGE_MARRIED', 'MARRIAGE_SINGLE', and 'MARRIAGE_OTHER' using `get_dummies()` function. 

These columns were added to the end of the dataframe to encode the marital status information in a binary format. 


<hr style="width:25%;margin-left:0;"> 

**Q5**. In the column 'EDUCATION', convert the values {0, 5, 6} to the value 4. 

(Total: 5 marks)  

In [40]:
# ---- provide your code here -----
df['EDUCATION'] = df['EDUCATION'].replace({0: 4, 5: 4, 6: 4})

,LIMIT_BAL,EDUCATION,AGE,PAY_0,PAY_2,PAY_3,PAY_4,PAY_5,PAY_6,BILL_AMT1,...,PAY_AMT2,PAY_AMT3,PAY_AMT4,PAY_AMT5,PAY_AMT6,default payment next month,SEX_FEMALE,MARRIAGE_MARRIED,MARRIAGE_SINGLE,MARRIAGE_OTHER
0,310000,3.0,32.0,0.0,0.0,0.0,0.0,0.0,0.0,172772.0,...,6000.0,4000.0,3000.0,1000.0,2000.0,0,False,True,False,False
1,10000,3.0,49.0,-1.0,-1.0,-2.0,-1.0,2.0,2.0,32.0,...,0.0,2828.0,0.0,182.0,0.0,1,True,True,False,False
2,50000,2.0,28.0,-1.0,-1.0,-1.0,0.0,-1.0,-1.0,430.0,...,46257.0,2200.0,1300.0,43987.0,1386.0,0,False,True,False,False
3,80000,3.0,52.0,2.0,2.0,3.0,3.0,3.0,2.0,36649.0,...,1600.0,1600.0,0.0,1600.0,1600.0,1,True,True,False,False
4,270000,1.0,34.0,1.0,2.0,0.0,0.0,2.0,0.0,20979.0,...,4000.0,2000.0,0.0,2000.0,2000.0,0,False,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10340,70000,3.0,53.0,2.0,2.0,2.0,2.0,2.0,2.0,54251.0,...,0.0,4400.0,2300.0,2300.0,2100.0,1,True,True,False,False
10341,220000,1.0,42.0,2.0,2.0,2.0,2.0,2.0,2.0,2500.0,...,0.0,0.0,0.0,0.0,0.0,1,True,True,False,False
10342,30000,2.0,51.0,5.0,4.0,3.0,2.0,3.0,2.0,26551.0,...,0.0,2113.0,0.0,0.0,2191.0,1,False,False,True,False
10343,20000,2.0,45.0,2.0,0.0,0.0,0.0,0.0,2.0,13696.0,...,1574.0,1284.0,1591.0,0.0,644.0,1,True,False,True,False


---
---

**Problem 3** Preparing X and y arrays (Total Marks: 10)

**Q1**. 

- Create a numpy array `y` from the first 7,500 observations of 'payment_default' column from `df` (2.5 marks)   
- Create a numpy array `X`  from the first 7,500 observations of all the remaining variables in `df` (2.5 marks)   

(Total: 5 Marks)

In [78]:
# ---- provide your code here -----
y = df.loc[0:7499, 'default payment next month'].values
X = df.loc[0:7499, df.columns != 'default payment next month'].values

<hr style="width:25%;margin-left:0;"> 

**Q2**. 

- Use an appropriate `sklearn` library we used in class to create `y_train`, `y_test`, `X_train` and `X_test` by splitting the data into 75% train and 25% test datasets (2.5 marks) 
    - Set random_state to 3 and stratify the subsamples so that train and test datasets have roughly equal proportions of the target's class labels 
- Standardise the data to mean zero and variance one using an approapriate `sklearn` library (2.5 marks)   

(Total: 5 marks) 

In [79]:
# ---- provide your code here -----
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 3, stratify = y)

sc = StandardScaler()
sc.fit(X_train)

X_train_scaled = sc.fit_transform(X_train)
X_test_scaled = sc.transform(X_test)

---
---

**Problem 4**. Support Vector Classifier and Accuracies (Total Marks: 30)


**Q1**. 

- Train a Support Vector Classifier on the standardised data (5 marks)
    - Use `rbf` kernel and set `random_state` to 24 (don't change any other parameters)
- Compute and print training and test dataset accuracies (5 marks)

(Total: 10 marks)   

In [48]:
# ---- provide your code here -----

from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

svm_clf = SVC(kernel='rbf', random_state=24)
svm_clf.fit(X_train_scaled, y_train)

# Compute training accuracy
train_preds = svm_clf.predict(X_train_scaled)
train_accuracy = accuracy_score(y_train, train_preds)

# Compute test accuracy
test_preds = svm_clf.predict(X_test_scaled)
test_accuracy = accuracy_score(y_test, test_preds)

# Print training and test dataset accuracies
print(f'SupportVector Train Accuracy: {train_accuracy:.4f}')
print(f'SupportVector Test Accuracy: {test_accuracy:.4f}')

SupportVector Train Accuracy: 0.8231
SupportVector Test Accuracy: 0.8144


<hr style="width:25%;margin-left:0;"> 

**Q2.**

- Extract 2 linear principal components from the standardised features using an appropriate `sklearn` library (5 marks)
- Train a Support Vector Classifier on the 2 principal components computed above (5 marks)   
    - Use `rbf` kernel and set `random_state` to 24 (don't change any other parameters)
- Compute and print training and test dataset accuracies (5 marks)


(Total: 15 marks)  

In [50]:
# ---- provide your code here -----
from sklearn.decomposition import PCA
from sklearn.svm import SVC

# Extract 2 linear principal components
pca = PCA(n_components=2)
X_train_pca = pca.fit_transform(X_train_scaled)
X_test_pca = pca.transform(X_test_scaled)

# Train Support Vector Classifier on the 2 principal components with RBF kernel
svm_clf_pca = SVC(kernel='rbf', random_state=24)
svm_clf_pca.fit(X_train_pca, y_train)

# Compute training accuracy
train_preds_pca = svm_clf_pca.predict(X_train_pca)
train_accuracy_pca = accuracy_score(y_train, train_preds_pca)

# Compute test accuracy
test_preds_pca = svm_clf_pca.predict(X_test_pca)
test_accuracy_pca = accuracy_score(y_test, test_preds_pca)

# Print training and test dataset accuracies
print(f'SupportVector Train Accuracy (PCA): {train_accuracy_pca:.4f}')
print(f'SupportVector Test Accuracy (PCA): {test_accuracy_pca:.4f}')

SupportVector Train Accuracy (PCA): 0.8012
SupportVector Test Accuracy (PCA): 0.8037


<hr style="width:25%;margin-left:0;"> 

**Q3**. 

- Comment on the suitability of the two classifiers to predict credit card defaults by commenting on (and comparing) the computed accuracies from the last two questions.


(Total: 5 marks)     


---- provide your text answer here ----

In Q1, Support Vector Classifier output on standardized features, we observe the training dataset accuracy of 0.8231 compared with the test accuracy of 0.8144. In Q2, Support Vector Classifier output when extracting 2 linear principal components, the accuracy is training accuracy is 0.8012 while the test accuracy is 0.8037.

Comparing two models based on the accuracies, the Support Vector Classifier on standardized features shows slightly greater accuracies than the Support Vector Classifier on 2 linear principal components. Thus, the Support Vector Classifier on standardized features is the optimal model for predicting credit card defaults dataset. 

Regarding to accuracy measures for the Support Vector Classifier on standardized features across the training and test datasets, the training accuracy is slightly greater, which means this model is a better predictor for the training dataset. However, in the Support Vector Classifier on 2 linear principal components, the test accuracy is slightly higher than the training accuracy. We can assume that it is because of the random fluctuation in the dataset which causes minor variation. Furthermore, the accuracies of the two models are similar across the training and test datasets, suggesting that overfitting is not a major problem for either model.

According to the accuracy outcomes and consistent performance across both training and test datasets, we can conclude that Support Vector Classifier on standardized features and the one on 2 linear principal components are both effective models in predicting credit card defaults. However, the one on standardized features has slightly higher accuracies, which makes it the preferable option of the two.

---
---

## Marking Criteria

To achieve a perfect score, your solutions must adhere to the criteria outlined below:

- Ensure that all numerical answers are accurate.
- Use the specific Python functions and libraries specified within the assignment instructions.
- For any written responses, provide accurate information, articulated in clear and complete sentences.
- Do not add extra cells beyond what is provided in the notebook.
- Do not print output with your code unless explicitly instructed to do so.
- Maintain a clean and organised notebook layout that is easy to follow.
    
---
---